velox/exec/Window.cpp

```diff
@@ -282,6 +314,131 @@ void Window::callResetPartition(vector_size_t partitionNumber) {
+          (isKPreceding ? -vector_size_t(offsets[i])
+                        : vector_size_t(offsets[i])) -
+          firstPartitionRow;
+    }
+  }
+}
+
+void Window::updateFrameBounds(
+    const vector_size_t functionIndex,
+    const bool isStartBound,
+    const vector_size_t startRow,
+    const vector_size_t numRows,
+    const vector_size_t* rawPeerStarts,
+    const vector_size_t* rawPeerEnds,
+    vector_size_t* rawFrameBounds) {
+  auto firstPartitionRow = partitionStartRows_[currentPartition_];
+  auto lastPartitionRow = partitionStartRows_[currentPartition_ + 1] - 1;
+  auto type = windowFrames_[functionIndex].type;
+  auto boundType = isStartBound ? windowFrames_[functionIndex].startType
+                                : windowFrames_[functionIndex].endType;
+  auto frameArg = isStartBound ? windowFrames_[functionIndex].start
+                               : windowFrames_[functionIndex].end;
+
+  switch (boundType) {
+    case core::WindowNode::BoundType::kUnboundedPreceding:
+      std::memset(rawFrameBounds, 0, numRows * sizeof(vector_size_t));
+      break;
+    case core::WindowNode::BoundType::kUnboundedFollowing:
+      std::fill_n(
+          rawFrameBounds, numRows, lastPartitionRow - firstPartitionRow);
+      break;
+    case core::WindowNode::BoundType::kCurrentRow: {
+      if (type == core::WindowNode::WindowType::kRange) {
+        const vector_size_t* rawPeerBuffer =
+            isStartBound ? rawPeerStarts : rawPeerEnds;
+        std::copy(rawPeerBuffer, rawPeerBuffer + numRows, rawFrameBounds);
+      } else {
+        // Fills the frameBound buffer with increasing value of row indices
+        // (corresponding to CURRENT ROW) from the startRow of the current
+        // output buffer. The startRow has to be adjusted relative to the
+        // partition start row.
+        std::iota(
+            rawFrameBounds,
+            rawFrameBounds + numRows,
+            startRow - firstPartitionRow);
+      }
+      break;
+    }
+    case core::WindowNode::BoundType::kPreceding: {
+      if (type == core::WindowNode::WindowType::kRows) {
+        updateKRowsFrameBounds(
+            true, frameArg.value(), startRow, numRows, rawFrameBounds);
+      } else {
+        VELOX_NYI("k preceding frame is only supported in ROWS mode");
+      }
+      break;
+    }
+    case core::WindowNode::BoundType::kFollowing: {
+      if (type == core::WindowNode::WindowType::kRows) {
+        updateKRowsFrameBounds(
+            false, frameArg.value(), startRow, numRows, rawFrameBounds);
+      } else {
+        VELOX_NYI("k following frame is only supported in ROWS mode");
+      }
+      break;
+    }
+    default:
+      VELOX_USER_FAIL("Invalid frame bound type");
+  }
+}
+
+namespace {
+void fixOutOfBoundFrames(
+    const vector_size_t& lastRow,
+    const vector_size_t& numRows,
+    vector_size_t* frameStart,
+    vector_size_t* frameEnd) {
+  vector_size_t* frameStartIndex = nullptr;
```

Comment:
Working with raw pointers like this makes the code hard to read. Consider refactoring.

```
void adjustKRowsFrameBounds(
    vector_size_t lastRow,
    vector_size_t numRows,
    vector_size_t* rawFrameStarts,
    vector_size_t* rawFrameEnds) {
  for (auto i = 0; i < numRows; i++) {
    auto frameStartIndex = rawFrameStarts[i];
    auto frameEndIndex = rawFrameEnds[i];
    // Clamp frameStart and frameEnd to the first and last row of the partition
    // if required.
    if (frameStartIndex <= frameEndIndex) {
      rawFrameStarts[i] = std::max(frameStartIndex, 0);
      rawFrameEnds[i] = std::min(frameEndIndex, lastRow);
    } else {
      VELOX_NYI("Empty frames are currently not supported");
    }
  }
}
```